### PR TITLE
Allow descriptions for payment fields

### PR DESCRIPTION
**Pull request by mattwire**
Overview
----------------------------------------
Allow a description to be specified for payment form fields.

Before
----------------------------------------
It is not possible to add a description for any of the payment fields.

After
----------------------------------------
A description can be added for any field added by `getPaymentFormFieldsMetadata()`
payment_descriptions

Technical Details
----------------------------------------
This adds an array of descriptions that is assigned to the template and can be used to display descriptions for each form element.  If the description is not specified there is no change.

Comments
----------------------------------------
Example usage in org.civicrm.smartdebit here: https://github.com/mattwire/org.civicrm.smartdebit/commit/931fbafde49d6b25c7e797e57aba8c1023d31235

### CODE REVIEW

**Comment by davejenx** [comment]
Applied as patch to 5.4.0, working well here with current org.civicrm.smartdebit master, thanks!

**Comment by mattwire** [comment]
@KarinG Is this something that could be of interest to you at all? That you might have time to review :-)  Works on webform as well.

File: templates/CRM/Core/BillingBlock.tpl [hunk 1]
```diff
@@ -41,7 +41,8 @@
             </div>
             <div class="content">
                 {$form.$paymentField.html}
-              {if $paymentField == 'cvv2'}{* @todo move to form assignment*}
+              {if $paymentFieldDescriptions.$name}<div class="description">{$paymentFieldDescriptions.$name}</div>
```
Comment (pradpnayak) [inline templates/CRM/Core/BillingBlock.tpl]: @mattwire Can you move the <div> section next for better styling?

Comment (mattwire) [reply templates/CRM/Core/BillingBlock.tpl]: @pradpnayak Yes, I've changed it - is that ok?

File: CRM/Core/Payment/Form.php [hunk 2]
```diff
@@ -129,8 +129,10 @@ protected static function addCommonFields(&$form, $paymentFields) {
       // This will cause the fields to be marked as required - but it is up to the payment processor to
       // validate it.
       $requiredPaymentFields[$field['name']] = $field['is_required'];
+      $paymentFieldDescriptions[$field['name']] = $field['description'];
     }
 
+    $form->assign('paymentFieldDescriptions', $paymentFieldDescriptions);
```
Comment (pradpnayak) [inline CRM/Core/Payment/Form.php]: Can we change this to
```
if (!empty($field['description'])) {
  $paymentFieldDescriptions[$field['name']] = $field['description'];
}
```

I feel there will be notice error when description is not present in $field.

File: CRM/Custom/Form/Field.php [hunk 3]
```diff
@@ -319,15 +319,21 @@ public function buildQuickForm() {
...
+    else {
+      $optionGroupParams = [
+        'is_reserved' => 0,
+        'is_active' => 1,
+        'options' => ['limit' => 0, 'sort' => "title ASC"],
+        'return' => ['title'],
+      ];
+    }
+
     // Get all custom (is_reserved=0) option groups
```
Comment (pradpnayak) [inline CRM/Custom/Form/Field.php]: Is this change required?

File: CRM/Core/Payment/Form.php [hunk 4]
```diff
@@ -129,8 +129,10 @@ protected static function addCommonFields(&$form, $paymentFields) {
       // This will cause the fields to be marked as required - but it is up to the payment processor to
       // validate it.
       $requiredPaymentFields[$field['name']] = $field['is_required'];
+      $paymentFieldDescriptions[$field['name']] = $field['description'];
```
Comment (pradpnayak) [inline CRM/Core/Payment/Form.php]: Can we change this to
```
if (!empty($field['description'])) {
  $paymentFieldDescriptions[$field['name']] = $field['description'];
}
```

I feel there will be notice error when description is not present in $field.

File: CRM/Custom/Form/Field.php (hunk 3 above)
Comment (mattwire) [reply CRM/Custom/Form/Field.php]: No, extra commit sneaked in

**Comment by mattwire** [comment]
test this please

File: CRM/Custom/Form/Field.php (hunk 3 above)
Comment (pradpnayak) [reply CRM/Custom/Form/Field.php]: I think they are back again ;)

Comment (mattwire) [reply CRM/Custom/Form/Field.php]: Gone again..

**Comment by eileenmcnaughton** [comment]
@mattwire this makes sense but I think we should assign the paymentfield metadata array to the template rather than an array specific to descriptions. Otherwise next time we want some extra info we have to define a new assign

**Comment by pradpnayak** [comment]
@mattwire would you like to update the code as per Eileen's [comment](https://github.com/civicrm/civicrm-core/pull/12721#issuecomment-416075542)?

**Comment by mattwire** [comment]
test this please

**Comment by pradpnayak** [comment]
retest this please

**Comment by pradpnayak** [comment]
Hi Matt,

I tested on current state for SEPA payment processor and it worked. Description fields are shown below payment fields.

**Before:**
before


**After:**
after


Thank you so much for the improvement!

**Comment by eileenmcnaughton** [comment]
test this please

**Comment by eileenmcnaughton** [comment]
test fails unrelated
